WIP: Documenting the `SharedTerm` module exports

ChrisEPhifer opened the pull request:
I've started adding some basic documentation to the functions exported from this module (per https://github.com/GaloisInc/saw-script/issues/387). Not going too crazy with it; would appreciate any feedback/requests for things to include in these docstrings.

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -745,14 +747,17 @@ reducePi sc t arg = do
     _ ->
       fail $ unlines ["reducePi: not a Pi term", showTerm t']
 
+-- | Copmute the type of a global variable.
```
typo

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -745,14 +747,17 @@ reducePi sc t arg = do
 scTypeOfDataType sc ident =
   dtType <$> scRequireDataType sc ident
 
+-- | Compute the type of an element of the @llvm.global_ctors@ array.
```
This operation has nothing to do with LLVM; rather, given the name of a constructor from a saw-core `data` declaration, this will compute its type.

ChrisEPhifer replied:
Ha, this is what I get for trying to piece together what everything is for by chasing names through files :)

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1077,62 +1084,88 @@ scDefTerm sc d = scGlobalDef sc (defIdent d)
 scNat :: SharedContext -> Natural -> IO Term
 scNat sc n = scFlatTermF sc (NatLit n)
 
+-- | Create a literal term from a string.
```
It might be nice to mention the saw-core type of the result: This makes terms of saw-core type `String` (not, for example, a vector of 8-bit chars)

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1077,62 +1084,88 @@ scDefTerm sc d = scGlobalDef sc (defIdent d)
 scString :: SharedContext -> String -> IO Term
 scString sc s = scFlatTermF sc (StringLit s)
 
+-- | Create a term representing the global type of strings.
```
Might be more precise to say "the primitive saw-core type `String`" (i.e. `@String@` in haddock syntax)

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1077,62 +1084,88 @@ scDefTerm sc d = scGlobalDef sc (defIdent d)
 scRecordSelect sc t fname = scFlatTermF sc (RecordProj t fname)
 
+-- | Create a term representing the type of a record from a list associating
+-- field names (as @String@s) and types (as @Term@s).
```
It'd be nice to mention that the ordering of fields doesn't matter for record types (see the docstring for the `RecordType` constructor of the `FlatTermF` datatype)

brianhuffman replied:
Also, you might consider using `'String'` and `'Term'` instead of `@String@` and `@Term@`, so that haddock will add hyperlinks (probably a good idea to do this consistently throughout). Note that you can also put single quotes inside `@` to hyperlink individual identifiers, e.g. `@'scRecordSelect' sc m@.

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1077,62 +1084,88 @@ scDefTerm sc d = scGlobalDef sc (defIdent d)
 scTupleType _ [t] = return t
 scTupleType sc (t : ts) = scPairType sc t =<< scTupleType sc ts
 
+-- | Create a term giving the left projection of a term repesenting a pair.
```
typo "repesenting"

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1145,23 +1178,31 @@ scTupleSelector sc t i n
   | otherwise = fail "scTupleSelector: non-positive index"
 
+-- | Create a term representing the type of a function, given a parameter and
+-- result type (as terms).
```
Maybe say that this creates a non-dependent function type

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1145,23 +1178,31 @@ scTupleSelector sc t i n
                   scTermF sc (Pi "_" a b')
 
+-- | Create a term representing the type of a function, given a list of
+-- parameter types and a result type (as terms).
```
As above, maybe say that this creates an non-dependent n-ary function type

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1145,23 +1178,31 @@ scTupleSelector sc t i n
 scFunAll sc argTypes resultType = foldrM (scFun sc) resultType argTypes
 
+-- | Create a lambda term from a parameter name, parameter type (as a term),
+-- and a body.
```
We should say something about conventions for de Bruijn indices.

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1145,23 +1178,31 @@ scTupleSelector sc t i n
 scLambda sc varname ty body = scTermF sc (Lambda varname ty body)
 
+-- | Create a lambda term of multiple arguments (curried) from a list
+-- associating parameter names to types (as terms) and a body.
```
As above, we should mention conventions for de Bruijn indices. Index 0 in the body refers to the last argument in the list, and index `n-1` refers to the first.

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1170,20 +1211,25 @@ scLambdaList _ [] rhs = return rhs
   scLambda sc nm tp =<< scLambdaList sc r rhs
 
+-- | Create a (possibly dependent) function given a parameter name, parameter
+-- type (as a term), and a body.
```
Again, should mention index convention: Index 0 is bound in the "body" term. We should probably also add docstrings to the individual `Term` arguments

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1254,52 +1302,66 @@ scVectorReduced sc ety xs
 scNatType sc = scFlatTermF sc preludeNatType
 
+-- | Create a term representing a vector type, from a term giving the length
+-- and a term giving the element type.
```
Per-argument docstrings would be nice (in general I like having them anywhere there are multiple arguments of the same type that mean different things)

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1254,52 +1302,66 @@ scVectorReduced sc ety xs
 scBvForall sc w f = scGlobalApply sc "Prelude.bvForall" [w, f]
 
--- ite :: (a :: sort 1) -> Bool -> a -> a -> a;
+-- | ite :: (a :: sort 1) -> Bool -> a -> a -> a;
```
For saw-core type signatures in docstrings, we should format them like `@foo :: ty@`

brianhuffman commented:
Thanks @ChrisEPhifer for working on this!

There are quite a few functions in this module that just apply some specific saw-core function to some arguments. We should try to document all of these using exactly the same style. Some such functions now have a docstring with the saw-core type signature of the constant, and I think that's generally useful. So an idea would be to have a short prose sentence, followed by the type signature in constant-width font on a new line; what do you think?

ChrisEPhifer commented:
Thanks for all your feedback and catching those typos @brianhuffman!

I think your idea for consistent style of those "call a specific function" cases is great. To make sure I'm clear on what you're suggesting, it would look something like:

```haskell
-- | <Prose description of function>
-- @foo :: <type signature>@
foo :: ...
foo ... = ...
```

brianhuffman commented:
Yeah, that looks like what I had in mind. But I was thinking that for maximum readability, we should ensure that there is actually an explicit line break before the type signature in the generated html. I'll check the haddock documentation to see how best to do that.

brianhuffman commented:
Ok, I think you should format the saw-core type signatures as haddock "code blocks" (see section 8.3.4 of the haddock docs  https://www.haskell.org/haddock/doc/html/ch03s08.html#idm140354810780208). A code block should be typeset on its own line(s) in html.

There are two ways to typeset a code block in haddock; the most concise way is with "bird tracks", where you just put a `>` at the start of the line:
```
-- | This is the prose description of function 'scFoo'.
--
-- > foo : (a : sort 0) -> Bool -> a -> a;
```
I'm not 100% sure whether the blank line before the code block is necessary, so maybe you try it and check the html output just to be sure.

ChrisEPhifer commented:
OK @brianhuffman, I think I addressed everything you brought up; I canonicalized the primitive term-building documentation, added some notes about the conventions with deBruijn indices, and documented (most) of the cases where argument types were duplicated (I didn't bother for the primitive application stuff, since the saw-core type makes it clear, I think.)

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1254,145 +1334,248 @@ scVectorReduced sc ety xs
--- primitive intAdd/intSub/intMul/intDiv/intMod :: Integer -> Integer -> Integer;
+-- | Create terms applying the (binary) integer primitives.
+--
+-- > intAdd/intSub/intMul/intDiv/intMod : Integer -> Integer -> Integer;
```
Does this actually generate reasonable html documentation? It might be preferable to split this into separate type signatures for each function.

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1403,89 +1586,125 @@ scIntMod sc x y = scGlobalApply sc "Prelude.intMod" [x, y]
--- primitive intNeg/intAbs :: Integer -> Integer;
+-- | Create terms applying the (unary) integer primitives.
+--
+-- > intNeg/intAbs : Integer -> Integer;
```
Same comment as above: The generated docs may be improved by splitting into separate type signatures.

brianhuffman commented on saw-core/src/Verifier/SAW/SharedTerm.hs:
```diff
@@ -1403,89 +1586,125 @@ scIntMod sc x y = scGlobalApply sc "Prelude.intMod" [x, y]
   scGlobalApply sc "Prelude.bvNat" [x, y]
 
+-- TODO: This doesn't appear to be used anywhere, and "FinVal" doesn't appear
+-- in Prelude.sawcore... can this be deleted?
```
If there is no `FinVal` in `Prelude.sawcore`, then you're right: We should get rid of it. Perhaps that's out of scope for this PR, though?

brianhuffman commented:
It looks like there are quite a few examples in this module of multiple top-level functions sharing a single type signature declaration. I can't remember exactly what haddock does with this; I suspect that it just duplicates the same docstring for each. (We should check to make sure.) I know that making separate type signatures for all these might seem like a lot of boilerplate, but if it makes the generated docs better then it's probably worth it.

It's really unfortunate that Template Haskell doesn't support haddock comments; if it did, then we could auto-generate all these term-constructing functions, and generate haddock strings that include the saw-core signatures, and we could even add support for doc-strings in the saw-core files, and include those in the haddocks too. Oh, well, maybe in another GHC release or two...

Anyway, another thing to look at while you're working on haddocks is in the export list: You might be able to make the documentation easier to use by rearranging the order of the export list and adding or rearranging subsection headings. Just arrange it however feels most useful to you, as someone trying to look up things in the documentation.

ChrisEPhifer commented:
You called it; should have checked to see what happens with docstrings applied to multiple type declarations - they're just duplicated for each function. I'll split them into more specific docs, that's no problem at all (more of an indication that I should have tried this out earlier than anything else.) I do wish Template Haskell could help us out here, though, that would be awesome.

I'll also add some organization via the export list as you recommended; forgot that was a thing :)

ChrisEPhifer commented:
@brianhuffman Sorry for spam if you already got a notification that I requested another review; just wanted to note that I separated the documentation as you recommended and added a bit more granularity to the section headers.
